presto-hive/src/main/java/com/facebook/presto/hive/util/FinishableAsyncQueue.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@ThreadSafe
+public class FinishableAsyncQueue<T>
+{
+    private static final CompletableFuture<?> COMPLETED_FUTURE_OF_NULL = CompletableFuture.completedFuture(null);
+
+    private final AsyncQueue<T> queue;
+    private final AtomicBoolean finished = new AtomicBoolean(false);
+
+    private final CompletableFuture<List<T>> completedFutureOfEmptyList = CompletableFuture.completedFuture(ImmutableList.of());
+
+    public FinishableAsyncQueue(int softMaxQueueSize)
+    {
+        queue = new AsyncQueue<>(softMaxQueueSize);
+    }
+
+    public int size()
+    {
+        return queue.size();
+    }
+
+    /**
+     * After calling this function, subsequent calls to {@link #offer(Object)} will be ignored. In addition,
+     * all futures returned previously will be immediately completed. Subsequent calls to {@link #offer(Object)}
+     * and {@link #getBatchAsync(int)} will always return a completed future. Ongoing calls to
+     * {@link #offer(Object)} may or may not be ignored.
+     */
+    public void finish()
+    {
+        if (finished.compareAndSet(false, true)) {
+            // It must be guaranteed that no one will become blocked once execution hits this point.
+            queue.signalAll();
+        }
+    }
+
+    public boolean isFinished()
+    {
+        // Once finished is set, size will never increase.
```

Comment:
I saw the point you are making. When I wrote the comment, I somehow wrote it in the wrong way.

Does the following look right?

It doesn't affect correctness because it's okay to ignore ongoing `offer`s when `finish` is called.

Once finish is called, `size` is at least the number of elements that have been completely put into the queue (`offer` has returned). For these elements, they must be among the first `size` elements in the queue. The first time `size` becomes 0 after `finish` is set, at least `size` elements have been removed, including all elements that have been completely put into the queue. Therefore, it is safe to return true when `size` is 0 and `finish` is set. Note however, isFinished may return `false` after it had returned `true`.